File: ios/Classes/Methods/PluginMethods.swift
import Flutter
import TwilioConversationsClient

class PluginMethods: NSObject, TWCONPluginApi {
    let TAG = "PluginMethods"

    func debugEnableNative(_ enableNative: NSNumber, enableSdk: NSNumber, error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        SwiftTwilioConversationsPlugin.nativeDebug = enableNative.boolValue
        if enableSdk.boolValue {
            TwilioConversationsClient.setLogLevel(TCHLogLevel.debug)
        } else {
            TwilioConversationsClient.setLogLevel(TCHLogLevel.warning)
        }
    }
    
    func createJwtToken(_ jwtToken: String?, completion: @escaping (TWCONConversationClientData?, FlutterError?) -> Void) {
        guard let jwtToken = jwtToken else {
            return completion(nil, FlutterError(code: "MISSING_PARAMS", message: "Missing 'token' parameter", details: nil))
        }
        debug("create => jwtToken: \(jwtToken)")
        
        let properties = TwilioConversationsClientProperties()
        //TODO add region to properties
                
        TwilioConversationsClient.conversationsClient(
            withToken: jwtToken,
            properties: properties,
            delegate: SwiftTwilioConversationsPlugin.clientListener,
            completion: { (result: TCHResult, conversationsClient: TwilioConversationsClient?) in
                if result.isSuccessful {
                    SwiftTwilioConversationsPlugin.debug("SwiftTwilioConversationsPlugin.create => ConversationsClient.create onSuccess: myIdentity is '\(conversationsClient?.user?.identity ?? "unknown")'")
                    conversationsClient?.delegate = SwiftTwilioConversationsPlugin.clientListener
                    SwiftTwilioConversationsPlugin.instance?.client = conversationsClient
                    let clientData = Mapper.conversationsClientToPigeon(conversationsClient)
                    completion(clientData, nil)
                } else {
                    SwiftTwilioConversationsPlugin.debug("SwiftTwilioConversationsPlugin.create => ConversationsClient.create onError: \(String(describing: result.error))")
                    completion(nil, FlutterError(code: "ERROR", message: "Error creating client, Error: \(result.error.debugDescription)", details: nil))
                }
        })
    }
    
    private func debug(_ msg: String) {
        SwiftTwilioConversationsPlugin.debug("\(TAG)::\(msg)")
    }
}

File: android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/PluginMethods.kt
package twilio.flutter.twilio_conversations.methods

import com.twilio.conversations.CallbackListener
import com.twilio.conversations.ConversationsClient
import com.twilio.conversations.ErrorInfo
import twilio.flutter.twilio_conversations.Api
import twilio.flutter.twilio_conversations.Mapper
import twilio.flutter.twilio_conversations.TwilioConversationsPlugin
import twilio.flutter.twilio_conversations.listeners.ClientListener

class PluginMethods : Api.PluginApi {
    private val TAG = "PluginMethods"

    override fun debug(enableNative: Boolean, enableSdk: Boolean) {
        if (enableSdk) {
            ConversationsClient.setLogLevel(ConversationsClient.LogLevel.DEBUG)
        } else {
            ConversationsClient.setLogLevel(ConversationsClient.LogLevel.ERROR)
        }

        TwilioConversationsPlugin.nativeDebug = enableNative
        return
    }

    override fun create(jwtToken: String, result: Api.Result<Api.ConversationClientData>) {
        debug("create => jwtToken: $jwtToken")
        val props = ConversationsClient.Properties.newBuilder().createProperties()

        ConversationsClient.create(TwilioConversationsPlugin.applicationContext, jwtToken, props, object :
            CallbackListener<ConversationsClient> {
            override fun onSuccess(conversationsClient: ConversationsClient) {
                TwilioConversationsPlugin.client = conversationsClient
                TwilioConversationsPlugin.clientListener = ClientListener()
                conversationsClient.addListener(TwilioConversationsPlugin.clientListener!!)
                val clientMap = Mapper.conversationsClientToPigeon(conversationsClient)
                result.success(clientMap)
            }

            override fun onError(errorInfo: ErrorInfo) {
                result.error(RuntimeException(errorInfo.message))
            }
        })
    }

    fun debug(message: String) {
        TwilioConversationsPlugin.debug("$TAG::$message")
    }
}
